sys/src/9/port/devramfs.c

```diff
@@ -0,0 +1,180 @@
+	ramroot->ramdir[0].length = 0;
+	ramroot->ramdir[0].perm = DMDIR|0777;
+
+	struct RamFile* file = (struct RamFile*)smalloc(sizeof(struct RamFile*) + strlen(readme));
+	file->length = strlen(readme);
+	memcpy(file->data, readme, file->length);
+	strcpy(ramroot->ramdir[1].name, "README");
+	ramroot->ramdir[1].qid = (Qid){(int64_t)file};
+	ramroot->ramdir[1].length = file->length;
+	ramroot->ramdir[1].perm = 0444;
+
+	struct Directory *subdir = (struct Directory *)smalloc(sizeof(struct Directory) + DIR_EXTEND_SIZE * sizeof(Dirtab));
+	strcpy(ramroot->ramdir[2].name, "subdir");
+	ramroot->ramdir[2].qid = (Qid){(int64_t)subdir, 0, QTDIR};
+	ramroot->ramdir[2].length = 0;
+        ramroot->ramdir[2].perm = DMDIR|0777;
+
+	subdir->inuse = 3;
+        subdir->allocated = DIR_EXTEND_SIZE;
+        strcpy(subdir->ramdir[0].name, "..");
+        subdir->ramdir[0].qid = (Qid){(int64_t)ramroot, 0,QTDIR};
+        subdir->ramdir[0].length = 0;
+        subdir->ramdir[0].perm = DMDIR|0777;
+
+        strcpy(subdir->ramdir[1].name, ".");
+        subdir->ramdir[1].qid = (Qid){(int64_t)subdir, 0,QTDIR};
+        subdir->ramdir[1].length = 0;
+        subdir->ramdir[1].perm = DMDIR|0777;
+
+
+        strcpy(subdir->ramdir[2].name, "README2");
+        subdir->ramdir[2].qid = (Qid){(int64_t)file};
+        subdir->ramdir[2].length = file->length;
+        subdir->ramdir[2].perm = 0444;
+
+}
+
+static void
+ramreset(void)
+{
+}
+
+static Chan*
+ramattach(char *spec)
+{
+        Chan *c;
+        char *buf;
+         
+        c = newchan();
+        mkqid(&c->qid, (int64_t)ramroot, 0, QTDIR);
+        c->dev = devtabget('@', 0);
+        if(spec == nil)
+                spec = "";
+        buf = smalloc(1+UTFmax+strlen(spec)+1);
+        sprint(buf, "#@%s", spec);
+        c->path = newpath(buf);
+	c->mode = 0777;
+        free(buf);
+        return c;
+}
+
+static Walkqid*
+ramwalk(Chan *c, Chan *nc, char **name, int nname)
+{
+	struct Directory *dir = (struct Directory *)c->qid.path;
+	return devwalk(c,  nc, name, nname, dir->ramdir, dir->inuse, devgen);
+}
+
+static int32_t
+ramstat(Chan *c, uint8_t *dp, int32_t n)
+{
+	struct Directory *dir = (struct Directory *)c->qid.path;
+	return devstat(c, dp, n, dir->ramdir, dir->inuse, devgen);
+}
+
+static Chan*
+ramopen(Chan *c, int omode)
+{
+	struct Directory *dir = (struct Directory *)c->qid.path;
+	return devopen(c, omode, dir->ramdir, dir->inuse, devgen);
+}
+
+/*
+ * sysremove() knows this is a nop
+ */
+static void
+ramclose(Chan* c)
+{
+}
+
+static int32_t
+ramread(Chan *c, void *buf, int32_t n, int64_t off)
+{
+	if (c->qid.type == QTDIR){
+		struct Directory *dir = (struct Directory*)c->qid.path;
+		return devdirread(c, buf, n, dir->ramdir, dir->inuse, devgen);
+	}
+	// Read file
+	struct RamFile *file = (void*)c->qid.path;
+	int filelen = file->length;
+	if (off > filelen){
+		return 0;
+	}
+	if (off + n > filelen){
+		n = filelen - off;
+	}
+	memcpy(buf, file->data, n);
```

Comment:
There are support functions to do all this. Take a look at, e.g., devarch or devproc.